Bluetooth: Initial conversion from bt_storage to the settings API

jhedberg opened the pull request:
This is an initial patch set to transition from the bt_storage API to a settings-based solution. It contains support for local values such as identity address and IRK, as well as remote pairing-related keys (the bt_keys struct). In follow-up PRs I'll work on adding support for per-device CCC as well as support for mesh storage.

Commit:
Bluetooth: Introduce skeleton for settings-based storage

Introduce a basic skeleton for peristent storage based on the settings
subsystem. Also enable support for this to the peripheral sample
application, so the new code gets exersized by CI. For now, the
implementation provides the same level support as the bt_storage API
ever did, i.e. for the identity address and the IRK.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Remove bt_storage API

The same functionality is now supported by the settings-based
solution, so remove bt_storage out of the way. There were stubs in
bt_storage to handle per-peer information (e.g. pairing keys) but this
was never actually implemented in full. The next step is to add this
support to the settings-based solution.

Leave the code for generating temporary IRK and identity address in
case BT_SETTINGS is not enabled. Also leave the code for using vendor
HCI to read the identity address, in which case the settings
implementation will not touch it.

Introduce a new bt_unpair() API to replace the removed
bt_storage_clear(), since the latter was actually doing more than just
storage management: it was also handling runtime storage of pairing
information. Later, the bt_unpair() implementation will be extended to
clear settings-based pairing storage.

There is one feature that the bt shell module looses: the ability to
give a specific identity address to the "init" command as a parameter.
We might look later in the future if this is really needed, and add a
separate API for this.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -0,0 +1,151 @@
+
+	BT_DBG("");
+
+	err = settings_subsys_init();
```
Note about this: I'm not 100% sure if the settings subsystem ensures that this API can be called multiple times (potentially by multiple subsystems). At least parts of it were protected by an "is_inited" boolean, but not all, I think. This might be good to fix in settings. Alternatively, we could move the requirement of calling settings_subsys_init() to the application side, i.e. before calling bt_enable(). That would make it consistent with the requirement of calling settings_load() on the application side as well. Thoughts?

jhedberg replied:
Before doing that however, we should fix settings so that settings_register() either cleanly fails or cleanly succeeds if called before settings_subsys_init() has been called. Currently that's not that case since settings_register() will always succeed, however settings_subsys_init() will clear the list of handlers if it's called later. I'd propose using static initialization of the handlers list, so that the sys_slist_init() call in settings_store_init() is not needed. That way it doesn't matter in which order things happen, and we can always safely call settings_register().

That said, I think I might still prefer for bt_enable() to fail if the application has not called settings_subsys_init(), which means that settings_register() should return failure in that case. Any thoughts @nvlsianpu ?

jhedberg replied:
One more thought: if we move the settings handlers to a dedicated ROM linker section, the settings_register() API will go away. That would change things a bit. One of the reasons I did settings_load() outside of bt_enable() is that with mesh the mesh settings handlers would only be registered upon bt_mesh_init() (which is called after bt_enable). However, if we have a dedicated linker sections, all handlers are already there from the very beginning, making it safer to call settings_load() at any time (it might still not be completely safe though since there might still be run-time init functions that some subsystems need to get called first)

jhedberg replied:
For the record: I've just pushed a third patch to this PR, showing how I'd do a ROM linker section for the Bluetooth-internal submodule handlers. The solution for the settings handlers would look quite similar.

nvlsianpu replied:
Agree, but have in mind that fcb test `tests/subsys/settings/fcb/src/settings_test_fcb.c'` need to be aligned as settings_register() is called without initaialization proir (any test_config_insert* testcase)

carlescufi replied:
Agree with the dependency on fixing `settings_register` before we can actually safely consider the initialization path. So should we put this PR on hold while we send a PR to fix the settings subsystem itself?

nvlsianpu replied:
Think not necessarily hood on this - as can be implemented properly even now.

Commit:
Bluetooth: settings: Add support for per-submodule handlers

Add a new linker section for a list of submodule settings handlers,
and iterate the list from the various settings callbacks.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1271,6 +1269,53 @@ static int set_flow_control(void)
+	return 0;
+}
+
+int bt_unpair(const bt_addr_le_t *addr)
```
Where exactly? Note that bt_unpair() is essentially the old bt_storage_clear() with the bt_storage->clear() called removed and moved to a new position in hci_core.c that's inside a `#if defined(CONFIG_BT_CONN)` section. Currently the API does not do any persistent clearing since there's no support for persistent pairing storage. As soon as it's added the API will do it, i.e. it'll then clear both runtime and persistent info, whereas currently it only clears runtime info (which is the only info we have for now).

carlescufi replied:
Of course, I failed to see that here. Disregard.

Commit:
Bluetooth: Add support for persistent pairing keys storage

Integrate the bt_keys submodule with bt_settings. Add a new
bt_keys_store() API to write keys to flash, and extend the existing
bt_keys_clear() to remove the keys from flash.

Along with this, add some helpers for genrating settings key values
containing a bluetooth address, as well as for decoding them to get
the binary bt_addr_le_t.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Compress bt_keys struct

There's a bit of unnecessary space in the bt_keys struct. Re-design
some fields for a more compact format, which is particularly helpful
now that the struct gets stored as-is to flash through the settings
API.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
I've now also pushed patches to add support for pairing keys storage. I also discovered that I was badly overflowing the Bluetooth RX stack as well as the system workqueue stack due to the settings API calls. So both of those stacks are now 2k by default (I've updated my first patch to reflect this).

Commit:
Bluetooth: shell: Add settings support

Enable settings and increase the system workqueue size to deal with
the stack usage. This also makes it possible to test unpairing support
with the shell's 'clear' command.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

dbkinder reviewed (approved):
Doc change LGTM, thanks.

jhedberg commented:
Any other comments on this PR? The "who should select MPU_ALLOW_FLASH_WRITE" is still an open issue (perhaps it should be SETTINGS rather than BT_SETTINGS), but I think that can be fixed up in later pull requests as well. It'd be good to get this merged since I'm starting to have other related patches queued up locally already (e.g. for mesh). From github policy perspective the review from @dbkinder is sufficient for merging, but it's only about the documentation, i.e. it'd be good to get a +1 from a functional perspective as well.
